velox/common/caching/SsdCache.cpp

```diff
@@ -0,0 +1,505 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/caching/SsdCache.h"
+#include <folly/executors/IOThreadPoolExecutor.h>
+#include <folly/portability/SysUio.h>
+#include <iostream>
+#include "velox/common/caching/FileIds.h"
+#include "velox/common/caching/GroupTracker.h"
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <numeric>
+
+DEFINE_bool(ssd_odirect, true, "use O_DIRECT for SSD cache IO");
+DEFINE_bool(verify_ssd_write, false, "Read back data after writing to SSD");
+
+namespace facebook::velox::cache {
+
+SsdPin::SsdPin(SsdFile& file, SsdRun run) : file_(&file), run_(run) {
+  file_->checkPinned(run_.offset());
+}
+
+SsdPin::~SsdPin() {
+  if (file_) {
+    file_->unpinRegion(run_.offset());
+  }
+}
+
+void SsdPin::operator=(SsdPin&& other) {
+  if (file_) {
+    file_->unpinRegion(run_.offset());
+  }
+  file_ = other.file_;
+  other.file_ = nullptr;
+  run_ = other.run_;
+}
+
+SsdFile::SsdFile(
+    const std::string& filename,
+    SsdCache& cache,
+    int32_t ordinal,
+    int32_t maxRegions)
+    : cache_(cache),
+      ordinal_(ordinal),
+      maxRegions_(maxRegions),
+      filename_(filename) {
+  int32_t oDirect = 0;
+#ifdef linux
+  oDirect = FLAGS_ssd_odirect ? O_DIRECT : 0;
+#endif
+  fd_ = open(filename.c_str(), O_CREAT | O_RDWR | oDirect, S_IRUSR | S_IWUSR);
+  if (fd_ < 0) {
+    LOG(ERROR) << "Cannot open or create " << filename << " error " << errno;
+    exit(1);
+  }
+  readFile_ = std::make_unique<LocalReadFile>(fd_);
+  uint64_t size = lseek(fd_, 0, SEEK_END);
+  numRegions_ = size / kRegionSize;
+  if (size % kRegionSize > 0) {
+    ftruncate(fd_, numRegions_ * kRegionSize);
+  }
+  // The existing regions in the file are writable.
+  for (auto i = 0; i < numRegions_; ++i) {
+    writableRegions_.push_back(i);
+  }
+  fileSize_ = numRegions_ * kRegionSize;
```

Comment:
perhaps, initialize fileSize_ right after numRegions and simplify the truncate logic above:

```
if (size > fileSize_) {
   ftruncate(fd_, fileSize_);
}
```